llumnix/backends/vllm/worker.py

```diff
@@ -11,51 +11,69 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
-from typing import List
+from typing import Dict, List
+import math
+import os
 import ray
 import torch
 
 from ray.util.scheduling_strategies import PlacementGroupSchedulingStrategy, NodeAffinitySchedulingStrategy
 from vllm.utils import is_pin_memory_available
 from vllm.worker.worker import Worker
+from vllm.config import CacheConfig, ModelConfig, ParallelConfig
+from vllm.worker.cache_engine import CacheEngine
 
 from llumnix.logger import init_logger
 from llumnix.backends.vllm.utils import _sample_with_torch
 
+from llumnix.backends.vllm.migrate_backend import get_migrate_backend
+from llumnix.config import MigrationConfig
+
 logger = init_logger(__name__)
 NUMPY_SUPPORT_DTYPES = [torch.float32, torch.float16]
 
-@ray.remote(num_cpus=0)
-class RecvActor:
-    def recv_cpu_cache(self, src_worker_handle, src_blocks, is_driver_worker):
-        """
-        Args:
-            src_worker_handle: src worker actor handle
-            blocks: block to send
-        """
-        try:
-            if is_driver_worker:
-                migration_cache = ray.get(src_worker_handle.execute_engine_method.remote("send_cpu_cache", src_blocks))
-            else:
-                migration_cache = ray.get(src_worker_handle.execute_method.remote("send_cpu_cache", src_blocks))
-        # pylint: disable=try-except-raise
-        except:
-            raise
-        return migration_cache
-
 class MigrationWorker(Worker):
     def __init__(self, *args, **kwargs) -> None:
         # replace sampler
         # pylint: disable=import-outside-toplevel
         import vllm.model_executor.layers.sampler
         vllm.model_executor.layers.sampler._sample_with_torch = _sample_with_torch
+
+        backend = os.environ.get("MIGRATE_BACKEND", "rpc")
+        migrate_size = int(os.environ.get("MIGRATE_CACHE_SIZE", 32))
+
+        parallel_config: ParallelConfig = kwargs["parallel_config"]
+        pp_or_tp_enabled = parallel_config.world_size > 1
+
+        if backend == "nccl" and (not pp_or_tp_enabled):
+            model_config: ModelConfig = kwargs["model_config"]
+            cache_config: CacheConfig = kwargs["cache_config"]
+            total_size = migrate_size * CacheEngine.get_cache_block_size(
+                cache_config, model_config, parallel_config) # for nccl backend gpu cache
+
+            device = torch.device(f"cuda:{kwargs['local_rank']}")
+            _, total_memory = torch.cuda.mem_get_info(device)
+            migrate_ratio = math.ceil(total_size / total_memory * 100) / 100
```

Comment:
total_size -> migration_memory might be better.

Reply:
How about dummy_cache_size ?